### PR TITLE
wallet: when a block is disconnected, update transactions that are no longer conflicted

### PR DESCRIPTION
**Pull request by ishaanam**
This implements a fix for #7315. Previously when a block was disconnected any transactions that were conflicting with transactions mined in that block were not updated to be marked as inactive. The fix implemented here is described on the [Bitcoin DevWiki](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Transaction-Conflict-Tracking#idea-refresh-conflicted). A test which tested the previous behavior has also been updated.

Second attempt at #17543

### CODE REVIEW

**Comment by ryanofsky** [comment]
Concept ACK on refreshing state of conflicted transactions in the wallet when a block is disconnected. I think there is a potential problem with the implementation though because it isn't recursively marking conflicted transactions as unconfirmed, only marking the top level transactions without recursing.

I think a good approach could be to steal from the previous implementation of this idea in commits b8b4592d985fa36dd0dddfd0e2bc6daf8df6e79c and 3c8bd6814ca88715d3d56994b10dc2c5f3e38e2d from #17543, which did recurse. That approach was also nice because it reused the existing MarkConflicted function in order to avoid code duplication. Commit ec5a89a26fa17e68b3e5928730e7d7c3e0370729 from that PR also added some more test coverage which could be used here.

**Comment by ishaanam** [comment]
@ryanofsky 
> I think there is a potential problem with the implementation though because it isn't recursively marking conflicted transactions as unconfirmed, only marking the top level transactions without recursing.

Yes, I had forgotten about this, thanks for pointing it out! 
> I think a good approach could be to steal from the previous implementation of this idea in commits https://github.com/bitcoin/bitcoin/commit/b8b4592d985fa36dd0dddfd0e2bc6daf8df6e79c and https://github.com/bitcoin/bitcoin/commit/3c8bd6814ca88715d3d56994b10dc2c5f3e38e2d from https://github.com/bitcoin/bitcoin/pull/17543, which did recurse.

I have used ideas from that PR so that the transaction state updating will recurse.
> Commit https://github.com/bitcoin/bitcoin/commit/ec5a89a26fa17e68b3e5928730e7d7c3e0370729 from that PR also added some more test coverage which could be used here.

I have cherry-picked that commit to this branch with a few adjustments.

File: test/functional/wallet_conflicts.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        self.log.info("Connect nodes 0 and 1, and ensure that the tx is effectively conflicted")
+        self.connect_nodes(0, 1)
+        self.sync_blocks([self.nodes[0], self.nodes[1]])
+        conflicted = self.nodes[0].gettransaction(conflicted_txid)
+        child_conflicted = self.nodes[0].gettransaction(child_conflicted_txid)
+        conflicting = self.nodes[0].gettransaction(conflicting_txid_A)
+        # Conflicted tx should have confirmations set to the confirmations of the most conflicting tx
+        assert_equal(conflicted["confirmations"], -conflicting["confirmations"])
+        # Child should inherit conflicted state from parent
+        assert_equal(child_conflicted["confirmations"], -conflicting["confirmations"])
```
Comment (glozow) [inline test/functional/wallet_conflicts.py:88]: We should also test who "wins" in this conflict contest and what the values are, i.e. conflicting A has 8 confirmations and conflicting B has 4 confirmations.

File: test/functional/wallet_conflicts.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        self.log.info("Disconnect nodes to broadcast conflicts on their respective chains")
+        self.disconnect_nodes(0, 1)
+        self.disconnect_nodes(2, 1)
+
+        self.log.info("Create a transaction with conflicted transactions spending outpoints A & B")
+        nA = next(tx_out["vout"] for tx_out in self.nodes[0].gettransaction(txid_conflict_from_1)["details"] if tx_out["amount"] == Decimal("10"))
+        nB = next(tx_out["vout"] for tx_out in self.nodes[0].gettransaction(txid_conflict_from_2)["details"] if tx_out["amount"] == Decimal("10"))
+        inputs_conflicted_tx = []
+        inputs_conflicted_tx.append({"txid": txid_conflict_from_1, "vout": nA})
+        inputs_conflicted_tx.append({"txid": txid_conflict_from_2, "vout": nB})
```
Comment (glozow) [inline test/functional/wallet_conflicts.py:41]: nit in 77a2e522d845eda270b2c03fd7e0a16bc4c00fbb, I think this is clearer (same with the other arrays of inputs and outputs)

```suggestion
        inputs_conflicted_tx = [{"txid": txid_conflict_from_1, "vout": nA}, {"txid": txid_conflict_from_2, "vout": nB}]
```

File: test/functional/wallet_conflicts.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        outputs_conflicted_tx = {}
+        outputs_conflicted_tx[self.nodes[0].getnewaddress()] = Decimal("19.99998")
+        outputs_conflicting_tx_A = {}
+        outputs_conflicting_tx_A[self.nodes[0].getnewaddress()] = Decimal("9.99998")
+        outputs_conflicting_tx_B = {}
+        outputs_conflicting_tx_B[self.nodes[0].getnewaddress()] = Decimal("9.99998")
+
+        conflicted = self.nodes[0].signrawtransactionwithwallet(self.nodes[0].createrawtransaction(inputs_conflicted_tx, outputs_conflicted_tx))
+        conflicting_tx_A = self.nodes[0].signrawtransactionwithwallet(self.nodes[0].createrawtransaction(inputs_conflicting_tx_A, outputs_conflicting_tx_A))
+        conflicting_tx_B = self.nodes[0].signrawtransactionwithwallet(self.nodes[0].createrawtransaction(inputs_conflicting_tx_B, outputs_conflicting_tx_B))
```
Comment (glozow) [inline test/functional/wallet_conflicts.py:56]: Sorry to bikeshed, but I think the naming is a bit hard to follow ("conflicted" vs "conflicting" don't immediately make sense to me, especially since both sets of transactions get conflicted at some point during the test). Might I suggest names that indicate which input(s) they spend, which node broadcasts them, and what "generation" they are? For example

`tx_AB_0_parent` instead of `conflicted`
`tx_A_1` instead of `conflicting_tx_A`
`tx_B_1` instead of `conflicting_tx_B`

`C` instead of `nA` for the 19.99998BTC output, because `nA` is already the name of the 10BTC output from one of the funding transactions.
`tx_C_0_child` instead of `child_conflicted`

File: test/functional/wallet_conflicts.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        assert_equal(conflicted["confirmations"], -conflicting["confirmations"])
+        # Child should inherit conflicted state from parent
+        assert_equal(child_conflicted["confirmations"], -conflicting["confirmations"])
+
+        # Node2 chain without conflicts
+        self.generate(self.nodes[2], 15, sync_fun=self.no_op)
+
+        # Connect node0 and node2 and wait reorg
+        self.connect_nodes(0, 2)
+        self.sync_blocks([self.nodes[0], self.nodes[2]])
```
Comment (glozow) [inline test/functional/wallet_conflicts.py:95]:
```suggestion
        self.sync_blocks()
```

**Review by glozow — COMMENTED** [review]
Concept ACK

There is a missing dash in "Co-authored-by:" in the commit messages

File: test/functional/wallet_conflicts.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python3
+# Copyright (c) 2023 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+"""Test conflicts tracking with multilple txn conflicting a conflicted tx."""
```
[the remark this answers is not in the document]
Comment (ishaanam) [reply test/functional/wallet_conflicts.py]: I've updated it to this, this is much more clear.

File: test/functional/wallet_conflicts.py (hunk 1 above)
Comment (ishaanam) [reply test/functional/wallet_conflicts.py]: I've added a check for this.

File: test/functional/wallet_conflicts.py (hunk 2 above)
Comment (ishaanam) [reply test/functional/wallet_conflicts.py]: Done

File: test/functional/wallet_conflicts.py (hunk 3 above)
Comment (ishaanam) [reply test/functional/wallet_conflicts.py]: I agree, the naming here was a bit confusing. It should be fixed now.

File: test/functional/wallet_conflicts.py (hunk 4 above)
Comment (ishaanam) [reply test/functional/wallet_conflicts.py]: Done both here and one other place.

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -1465,8 +1465,32 @@ void CWallet::blockDisconnected(const interfaces::BlockInfo& block)
     // future with a stickier abandoned state or even removing abandontransaction call.
     m_last_block_processed_height = block.height - 1;
     m_last_block_processed = *Assert(block.prev_hash);
+
+    const int& disconnect_height = block.height;
```
[the remark this answers is not in the document]
Comment (ishaanam) [reply src/wallet/wallet.cpp]: Fixed

**Comment by ishaanam** [comment]
Thanks for the review @glozow, I have addressed your comments and also made a modification to `RecursiveUpdateTxState` which will allow it to be more extensible.

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -1336,13 +1336,29 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
...
+            return true;
+        }
+        return false;
+    };
+
+    RecursiveUpdateTxState(hashTx, get_updated_state);
+
+}
+
+void CWallet::RecursiveUpdateTxState(const uint256& tx_hash, const GetUpdatedStateFn& get_updated_state) {
```
Comment (achow101) [inline src/wallet/wallet.cpp:1353]: In c3fe92175b6ac709e4fdfd4766253c75d7ffd184 "wallet: introduce generic recursive tx state updating function"

This could also be used by `AbandonTransaction`.

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -1351,11 +1367,7 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
...
         CWalletTx& wtx = it->second;
-        int currentconfirm = GetTxDepthInMainChain(wtx);
-        if (conflictconfirms < currentconfirm) {
-            // Block is 'more conflicted' than current confirm; update.
-            // Mark transaction as conflicted with this block.
-            wtx.m_state = TxStateConflicted{hashBlock, conflicting_height};
+        if (get_updated_state(wtx)) {
             wtx.MarkDirty();
             batch.WriteTx(wtx);
             // Iterate over all its outputs, and mark transactions in the wallet that spend them conflicted too
```
Comment (achow101) [inline src/wallet/wallet.cpp:1373]: In c3fe92175b6ac709e4fdfd4766253c75d7ffd184 "wallet: introduce generic recursive tx state updating function"

The comments should be updated to avoid mentioning specific transaction states.

File: src/wallet/wallet.cpp [hunk 9]
```diff
@@ -1336,13 +1313,29 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
...
+            return true;
+        }
+        return false;
+    };
+
+    RecursiveUpdateTxState(hashTx, get_updated_state);
+
+}
+
+void CWallet::RecursiveUpdateTxState(const uint256& tx_hash, const GetUpdatedStateFn& get_updated_state) {
```
[the remark this answers is not in the document]
Comment (ishaanam) [reply src/wallet/wallet.cpp]: Yes, I've changed `GetUpdatedStateFn` to `TryUpdatingStateFn`, and `get_updated_state` to `try_updating_state`.

**Review by furszy — COMMENTED** [review]
Haven't finished checking the latest changes but, if you like it, https://github.com/furszy/bitcoin-core/commit/8bb14c1eaf09b0bca9efa78e1a7a7a00c61c0a29 is all yours. During the review, the functional test wasn't clear enough for me so I spent a bit of time making it more friendly.

File: src/wallet/wallet.cpp [hunk 10]
```diff
@@ -1336,13 +1313,27 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
...
+        }
+        return false;
+    };
+
+    RecursiveUpdateTxState(hashTx, try_updating_state);
+
+}
+
+void CWallet::RecursiveUpdateTxState(const uint256& tx_hash, const TryUpdatingStateFn& try_updating_state) {
+    LOCK(cs_wallet);
```
Comment (furszy) [inline src/wallet/wallet.cpp:1329]: In f998cdbc:
This lock isn't needed. All callers have `cs_wallet` acquired. Could add the lock required annotation instead.

File: src/wallet/wallet.cpp [hunk 11]
```diff
@@ -1282,44 +1277,26 @@ bool CWallet::AbandonTransaction(const uint256& hashTx)
...
         assert(currentconfirm <= 0);
-        // if (currentconfirm < 0) {Tx and spends are already conflicted, no need to abandon}
+
         if (currentconfirm == 0 && !wtx.isAbandoned()) {
-            // If the orig tx was not in block/mempool, none of its spends can be in mempool
             assert(!wtx.InMempool());
             wtx.m_state = TxStateInactive{/*abandoned=*/true};
-            wtx.MarkDirty();
-            batch.WriteTx(wtx);
             NotifyTransactionChanged(wtx.GetHash(), CT_UPDATED);
```
Comment (furszy) [inline src/wallet/wallet.cpp:1287]: In f998cdbc:

Would be good to move this `NotifyTransactionChanged` to `RecursiveUpdateTxState` too. So `MarkConflicted` and the `BlockDisconnected` txes state changes are notified to the upper layers.

Comment (ishaanam) [reply src/wallet/wallet.cpp:1287]: I think that this is a good idea, but would out of the scope of this PR and would warrant more discussion, since then it would be notifying for tx state updates that it previously did not.

File: src/wallet/wallet.cpp (hunk 10 above)
Comment (ishaanam) [reply src/wallet/wallet.cpp]: Done

File: src/wallet/wallet.cpp [hunk 12]
```diff
@@ -1443,8 +1443,36 @@ void CWallet::blockDisconnected(const interfaces::BlockInfo& block)
...
+
+            std::pair<TxSpends::const_iterator, TxSpends::const_iterator> range = mapTxSpends.equal_range(tx_in.prevout);
+
+            // For all of the spends that conflict with this transaction
+            for (TxSpends::const_iterator _it = range.first; _it != range.second; ++_it) {
+                CWalletTx& wtx = mapWallet.find(_it->second)->second;
+
+                if (!wtx.isConflicted()) continue;
+
+                auto try_updating_state = [&](CWalletTx& wtx) {
```
[the remark this answers is not in the document]
Comment (ishaanam) [reply src/wallet/wallet.cpp]: > The lambda function's wtx shadows the function's wtx.

I have renamed the lambda function's `wtx`.
> Could just provide disconnect_height.

By this do you mean providing the `disconnect_height` of the block to `RecursiveUpdateTxState` or providing the wallet tx's `conflicting_block_height` to `try_updating_state` instead of the whole wallet tx by reference?

**Comment by ishaanam** [comment]
Thanks for the review @achow101 and @furszy!
> Haven't finished checking the latest changes but, if you like it, https://github.com/furszy/bitcoin-core/commit/8bb14c1eaf09b0bca9efa78e1a7a7a00c61c0a29 is all yours. During the review, the functional test wasn't clear enough for me so I spent a bit of time making it more friendly.

Thanks, I'll use these improvements here once I review them.

File: src/wallet/wallet.cpp (hunk 11 above)
Comment (ryanofsky) [reply src/wallet/wallet.cpp:1287]: In commit "wallet: introduce generic recursive tx state updating function" (f998cdbc689cc65ff0aca98ce63758c59786c1db)

> I think that this is a good idea, but would out of the scope of this PR and would warrant more discussion, since then it would be notifying for tx state updates that it previously did not.

Agree we should avoid changing behavior, but I think the current commit might also be changing behavior by notifying before calling MarkDirty and WriteTx instead of after.

I think the best thing to do would be to change TryUpdatingStateFn to return an enum instead of a bool like:

```c++
enum class TxUpdate { UNCHANGED, CHANGED, NOTIFY_CHANGED };
using TryUpdatingStateFn = std::function<TxUpdate(CWalletTx& wtx)>;
```

Returning CHANGED/UNCHANGED should make the code clearer anyway because it's less obvious what true and false mean.

File: src/wallet/wallet.cpp [hunk 13]
```diff
@@ -1336,13 +1313,27 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
     if (conflictconfirms >= 0)
         return;
 
-    // Do not flush the wallet here for performance reasons
```
Comment (ryanofsky) [inline src/wallet/wallet.cpp:1339]: In commit "wallet: introduce generic recursive tx state updating function" (f998cdbc689cc65ff0aca98ce63758c59786c1db)

This also seems like a useful comment. Would keep instead of deleting

File: src/wallet/wallet.cpp [hunk 14]
```diff
@@ -1336,13 +1313,27 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
...
+    auto try_updating_state = [&](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
+        if (conflictconfirms < GetTxDepthInMainChain(wtx)) {
+            wtx.m_state = TxStateConflicted{hashBlock, conflicting_height};
+            return true;
+        }
+        return false;
+    };
+
+    RecursiveUpdateTxState(hashTx, try_updating_state);
```
Comment (ryanofsky) [inline src/wallet/wallet.cpp:1324]: In commit "wallet: introduce generic recursive tx state updating function" (f998cdbc689cc65ff0aca98ce63758c59786c1db)

This code pretty opaque, would suggest copying the previous comment here "Iterate over all its outputs, and mark transactions in the wallet that spend them conflicted too" or something equivalent.

File: src/wallet/wallet.cpp [hunk 15]
```diff
@@ -1336,13 +1313,27 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
     if (conflictconfirms >= 0)
         return;
 
-    // Do not flush the wallet here for performance reasons
+    auto try_updating_state = [&](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
+        if (conflictconfirms < GetTxDepthInMainChain(wtx)) {
+            wtx.m_state = TxStateConflicted{hashBlock, conflicting_height};
```
Comment (ryanofsky) [inline src/wallet/wallet.cpp:1318]: In commit "wallet: introduce generic recursive tx state updating function" (f998cdbc689cc65ff0aca98ce63758c59786c1db)

Would suggest keeping the previous comment here instead of deleting it. It's not obvious what the check means or what's happening in the block:

```c++
// Block is 'more conflicted' than current confirm; update.
// Mark transaction as conflicted with this block.
```

File: src/wallet/wallet.cpp (hunk 11 above)
Comment (ishaanam) [reply src/wallet/wallet.cpp]: I have implemented this suggestion, I agree that it is much clearer now.

File: src/wallet/wallet.cpp [hunk 16]
```diff
@@ -1282,44 +1277,26 @@ bool CWallet::AbandonTransaction(const uint256& hashTx)
...
-        todo.erase(now);
-        done.insert(now);
-        auto it = mapWallet.find(now);
-        assert(it != mapWallet.end());
-        CWalletTx& wtx = it->second;
+    auto try_updating_state = [&](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
         int currentconfirm = GetTxDepthInMainChain(wtx);
-        // If the orig tx was not in block, none of its spends can be
         assert(currentconfirm <= 0);
-        // if (currentconfirm < 0) {Tx and spends are already conflicted, no need to abandon}
```
[the remark this answers is not in the document]
Comment (ishaanam) [reply src/wallet/wallet.cpp:1297]: I implemented your next suggestion about simplifying the code in the lambda, so the comments are a bit different now but I did add them back.

File: src/wallet/wallet.cpp [hunk 17]
```diff
@@ -1282,44 +1277,26 @@ bool CWallet::AbandonTransaction(const uint256& hashTx)
...
-
-    while (!todo.empty()) {
-        uint256 now = *todo.begin();
-        todo.erase(now);
-        done.insert(now);
-        auto it = mapWallet.find(now);
-        assert(it != mapWallet.end());
-        CWalletTx& wtx = it->second;
+    auto try_updating_state = [&](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
         int currentconfirm = GetTxDepthInMainChain(wtx);
```
[the remark this answers is not in the document]
Comment (ishaanam) [reply src/wallet/wallet.cpp]: This is simpler and it is a small change so I've implemented it in this commit.

File: src/wallet/wallet.cpp (hunk 13 above)
Comment (ishaanam) [reply src/wallet/wallet.cpp:1339]: I have added it back.

File: src/wallet/wallet.cpp (hunk 14 above)
Comment (ishaanam) [reply src/wallet/wallet.cpp:1326]: Done

File: src/wallet/wallet.cpp (hunk 15 above)
Comment (ishaanam) [reply src/wallet/wallet.cpp:1319]: Done

File: src/wallet/wallet.cpp [hunk 18]
```diff
@@ -1282,44 +1277,25 @@ bool CWallet::AbandonTransaction(const uint256& hashTx)
...
-        assert(it != mapWallet.end());
-        CWalletTx& wtx = it->second;
-        int currentconfirm = GetTxDepthInMainChain(wtx);
-        // If the orig tx was not in block, none of its spends can be
-        assert(currentconfirm <= 0);
-        // if (currentconfirm < 0) {Tx and spends are already conflicted, no need to abandon}
-        if (currentconfirm == 0 && !wtx.isAbandoned()) {
-            // If the orig tx was not in block/mempool, none of its spends can be in mempool
-            assert(!wtx.InMempool());
+    auto try_updating_state = [&](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
```
Comment (furszy) [inline src/wallet/wallet.cpp:1280]: Can remove the context ref.
```suggestion
    auto try_updating_state = [](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
```

File: src/wallet/wallet.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/wallet.cpp]: Yeah, nothing biggie, just:
```c++
auto try_updating_state = [disconnect_height](CWalletTx& tx) {
```
so only the required elements are provided to the lambda.

File: src/wallet/wallet.cpp (hunk 18 above)
Comment (ishaanam) [reply src/wallet/wallet.cpp]: Done

**Comment by ishaanam** [comment]
Rebased to fix failing CI.

File: test/functional/wallet_abandonconflict.py [hunk 19]
```diff
@@ -226,20 +226,16 @@ def run_test(self):
...
         # Verify that B and C's 10 BTC outputs are available for spending again because AB1 is now conflicted
+        assert_equal(alice.gettransaction(txAB1)["confirmations"], -1)
         newbalance = alice.getbalance()
         assert_equal(newbalance, balance + Decimal("20"))
         balance = newbalance
 
-        # There is currently a minor bug around this and so this test doesn't work.  See Issue #7315
-        # Invalidate the block with the double spend and B's 10 BTC output should no longer be available
-        # Don't think C's should either
+        # Invalidate the block with the double spend. B & C's 10 BTC outputs should no longer be available
```
Comment (furszy) [inline test/functional/wallet_abandonconflict.py:234]: Let me see if I'm following this:

`Tx_AB1` spends B, and `Tx_ABC2` (the child of `Tx_AB1`) spends C. So, when the double spend tx gets disconnected, `Tx_AB1` state changes to `inactive`, marking output B and C as spent again.

Which ends up with the txes stuck now right?, and the user need to manually re-submit them.

**Review by furszy — APPROVED** [review]
ACK 6f1d3948

File: test/functional/wallet_abandonconflict.py (hunk 19 above)
Comment (furszy) [reply test/functional/wallet_abandonconflict.py:234]: good. Not sure if there is any but would be nice to have coverage for the automatic re-submission. Would just need to mock the node time 24hs in the future and see if they get re-submitted.

**Comment by ishaanam** [comment]
I've fixed a bug which resulted in the conflicted tx not being marked as inactive if the conflicting tx was not in `mapTxSpends`.

File: test/functional/wallet_abandonconflict.py (hunk 19 above)
Comment (ishaanam) [reply test/functional/wallet_abandonconflict.py:234]: There are some tests for that here: https://github.com/bitcoin/bitcoin/blob/master/test/functional/wallet_resendwallettransactions.py. Or did you mean that a test for that should be added for this PR specifically?

**Review by rajarshimaitra — COMMENTED** [review]
tACK 89df7987c2f1eea42454c2b0efc31a924fbfd3a8.

Changes look very clean. We reviewed it in out weekly club and went over the new code addition. 

Ran the tests on the master branch and saw it failing `assert_equal(conflicted["confirmations"], 0)`. In the master conflicting transaction's state doesn't update after reorg.

**Comment by achow101** [comment]
ACK 89df7987c2f1eea42454c2b0efc31a924fbfd3a8

**Comment by furszy** [comment]
> I've fixed a bug which resulted in the conflicted tx not being marked as inactive if the conflicting tx was not in mapTxSpends.

Could you expand this?

Not sure if I follow the [latest push](https://github.com/bitcoin/bitcoin/compare/6f1d394896fcc7e2687b766418f9cd4e55b613b5..89df7987c2f1eea42454c2b0efc31a924fbfd3a8), where only the inequality was changed.
If `mapTxSpends.count(outpoint) == 1`, the outpoint was spent only once, so the wallet shouldn't contain any conflicting tx.

File: test/functional/wallet_abandonconflict.py (hunk 19 above)
Comment (furszy) [reply test/functional/wallet_abandonconflict.py:234]: yeah, great if that is already covered.
(sorry for the delayed answer)

**Comment by glozow** [comment]
> I've fixed a bug which resulted in the conflicted tx not being marked as inactive if the conflicting tx was not in mapTxSpends.

> If mapTxSpends.count(outpoint) == 1, the outpoint was spent only once, so the wallet shouldn't contain any conflicting tx.

I interpreted the bug as = we wouldn't mark a conflicted tx inactive if the block transaction was not in `mapTxSpends` (which would mean count == 1). Normally we'd expect them both to be there (count > 1).

**Review by glozow — COMMENTED** [review]
ACK 89df7987c2f1eea42454c2b0efc31a924fbfd3a8

**Comment by ishaanam** [comment]
>  If mapTxSpends.count(outpoint) == 1, the outpoint was spent only once, so the wallet shouldn't contain any conflicting tx.

This is not necessarily always true, because even though the conflicted wallet transaction must always have its spends added to `mapTxSpends`, this is not the case for all conflicting transactions. 

For example, let's say that Alice has a utxo. She creates a transaction, tx1, spending this utxo to Bob. `AddToSpends` will get called on Bob's node for this transaction, because Bob is the recipient. Alice then creates a conflicting transaction to tx1, called tx2, which sends the funds back to herself. However, Bob's node does not call `AddToSpends` on tx2, because Bob is not the recipient. If tx2 gets mined, but then that block is disconnected from our node, Bob's `mapTxSpends` will only show a single spend for Alice's utxo in `mapTxSpends`, even though there are technically two.

**Comment by furszy** [comment]
Ok nice, great. Thanks for the details.
We should code a test exercising exactly that behavior.

**Review by furszy — APPROVED** [review]
Tested ACK 89df7987

> For example, let's say that Alice has a utxo. She creates a transaction, tx1, spending this utxo to Bob. AddToSpends will get called on Bob's node for this transaction, because Bob is the recipient. Alice then creates a conflicting transaction to tx1, called tx2, which sends the funds back to herself. However, Bob's node does not call AddToSpends on tx2, because Bob is not the recipient. If tx2 gets mined, but then that block is disconnected from our node, Bob's mapTxSpends will only show a single spend for Alice's utxo in mapTxSpends, even though there are technically two.

Made a quick unit test and all good. Good finding.
Would be good to have a real functional test in the future.

**Comment by ishaanam** [comment]
> Would be good to have a real functional test in the future.

#27307 tests this behavior, which is how I found this bug.
